### PR TITLE
Add ability to see Zigbee information for a device

### PR DESCRIPTION
**Pull request by dmulcahey**
## Proposed change

Add dialog to expose the device signature for ZHA Zigbee devices. Do not merge without: https://github.com/home-assistant/core/pull/33612

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  public async showDialog(
+    params: ZHADeviceZigbeeInfoDialogParams
+  ): Promise<void> {
+    this._params = params;
+    const signature = JSON.parse(JSON.stringify(this._params.device.signature));
+    signature.manufacturer = this._params.device.manufacturer;
+    signature.model = this._params.device.model;
+    this._signature = JSON.stringify(signature, undefined, 2);
+    await this.updateComplete;
+    this._dialog.open();
```
Comment (zsarnett) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:43]: Not needed

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      >
+        ${this._error
+          ? html`
+              <div class="error">${this._error}</div>
+            `
+          : html`
+              <ha-code-editor
+                class="editor"
+                mode="yaml"
+                .value="${this._signature}"
```
Comment (zsarnett) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:70]:
```suggestion
                .value=${this._signature}
```

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+  public async showDialog(
+    params: ZHADeviceZigbeeInfoDialogParams
+  ): Promise<void> {
+    this._params = params;
+    const signature = JSON.parse(JSON.stringify(this._params.device.signature));
+    signature.manufacturer = this._params.device.manufacturer;
+    signature.model = this._params.device.model;
+    this._signature = JSON.stringify(signature, undefined, 2);
+    await this.updateComplete;
```
Comment (zsarnett) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:42]:
```ts
fireEvent(this._dialog as HTMLElement, "iron-resize");
```

resize after this?

File: src/panels/config/zha/zha-device-card.ts [hunk 4]
```diff
@@ -340,6 +342,24 @@ class ZHADeviceCard extends LitElement {
                           : ""}
                       `
                     : ""}
+                  ${this.device!.device_type !== "Coordinator"
+                    ? html`
+                        <mwc-button @click="${this._handleZigbeeInfoClicked}">
```
Comment (zsarnett) [inline src/panels/config/zha/zha-device-card.ts:347]:
```suggestion
                        <mwc-button @click=${this._handleZigbeeInfoClicked}>
```

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    });
+  }
+
+  protected render(): TemplateResult {
+    if (!this._params) {
+      return html``;
+    }
+
+    return html`
+      <ha-paper-dialog
```
Comment (bramkragten) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:62]: We should stop adding new `ha-paper-dialog`s and only add `ha-dialog`s (mwc)

Comment (bramkragten) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:62]: Will not block this now.

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  @property() private _error?: string;
+  @property() private _signature: any;
+  @query("ha-code-editor") private _codeEditor?: HaCodeEditor;
+  @query("ha-paper-dialog") private _dialog!: HaPaperDialog;
+
+  public async showDialog(
+    params: ZHADeviceZigbeeInfoDialogParams
+  ): Promise<void> {
+    this._params = params;
+    const signature = JSON.parse(JSON.stringify(this._params.device.signature));
```
Comment (bramkragten) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:40]: Why first stringify and then parse?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    signature.manufacturer = this._params.device.manufacturer;
+    signature.model = this._params.device.model;
+    this._signature = JSON.stringify(signature, undefined, 2);
```
Comment (bramkragten) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:43]: Should we show this as yaml?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        @opened-changed=${this._openedChanged}
+      >
+        ${this._error
+          ? html`
+              <div class="error">${this._error}</div>
+            `
+          : html`
+              <paper-dialog-scrollable>
+                <ha-code-editor
+                  mode="yaml"
```
Comment (bramkragten) [inline src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:75]: Signature is `json` should we add mode json? And a read-only mode? Or just put this in a `<pre>` block?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 6 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:40]: deep copy, is there a better way?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 7 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:43]: no

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 8 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:75]: Sure, how would I go about setting read-only mode up? Assuming just change the mode to json for the 1st part?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 5 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:62]: I can do it in this. I don't mind. I'll look into it.

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 8 above)
Comment (bramkragten) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:75]: I don't think we include a `json` mode for codemirror at the moment, but maybe `yaml` mode is fine?

<https://github.com/home-assistant/frontend/blob/903f92f94a40cdec52f196f581343fd09241ee8b/src/resources/codemirror.ts#L5-L6>

Read only should be added here if needed:
<https://github.com/home-assistant/frontend/blob/dev/src/components/ha-code-editor.ts#L119>

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 6 above)
Comment (bramkragten) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:40]: Why do you need a deep copy? You are not going to change it? (and yes, see <https://www.npmjs.com/package/deep-clone-simple>)

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 3 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:44]: it's in the block below?

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 5 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts]: Think i got it...

File: src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts (hunk 6 above)
Comment (dmulcahey) [reply src/dialogs/zha-device-zigbee-signature-dialog/dialog-zha-device-zigbee-info.ts:37]: um i change it on the next line...

**Review by bramkragten — APPROVED** [review]

**Pull request merged**